tests/jax/test_fused_attn.py

```diff
@@ -137,82 +195,27 @@ def set_input(self, b, s, h, d, *, attn_bias_type, attn_mask_type, dropout_proba
-    def test_sanity(self, b, s, h, d, attn_bias_type, attn_mask_type, dropout_probability, dtype,
-                    is_training, pad_ratio):
-
-        def grad_func(func, *args, **kwargs):
-            # Keep only valid result for the gradient
-            # fused_attn_max_512 output has shape (b, s, h, d)
-            valid_ret, _ = jnp.split(func(*args, **kwargs), (self.valid_len,), axis=1)
-            return jnp.mean(valid_ret, dtype=jnp.float32).astype(dtype)
-
-        self.set_input(b,
-                       s,
-                       h,
-                       d,
-                       attn_bias_type=attn_bias_type,
-                       attn_mask_type=attn_mask_type,
-                       dropout_probability=dropout_probability,
-                       dtype=dtype,
-                       is_training=is_training,
-                       pad_ratio=pad_ratio)
-
-        kwargs = {
-            'attn_bias_type': self.attn_bias_type,
-            'attn_mask_type': attn_mask_type,
-            'scaling_factor': self.scaling_factor,
-            'dropout_probability': self.dropout_probability,
-            'is_training': self.is_training
-        }
-
-        jitted_primitive = jit(
-            value_and_grad(
-                lambda qkv, bias, q_token, kv_token, dropout_rng: grad_func(
-                    customcall_self_fused_attn, qkv, bias, q_token, kv_token, dropout_rng, **kwargs
-                ), (0, 1)))
-
-        primitive_out, (primitive_dqkv,
-                        primitive_dbias) = jitted_primitive(self.qkv, self.bias, self.q_token,
-                                                            self.kv_token, self.dropout_rng)
-
-    @pytest.mark.parametrize('b, s, h, d', SELF_CASES)
-    @pytest.mark.parametrize('attn_bias_type', [AttnBiasType.NO_BIAS, AttnBiasType.POST_SCALE_BIAS])
-    @pytest.mark.parametrize('attn_mask_type',
-                             [AttnMaskType.PADDING_MASK, AttnMaskType.CAUSAL_MASK])
-    @pytest.mark.parametrize('dropout_probability', [0., 0.1])
-    @pytest.mark.parametrize('dtype', DTYPES)
-    @pytest.mark.parametrize('is_training', [True, False])
-    @pytest.mark.parametrize('pad_ratio', PAD_RATIO)
-    def test_forward(self, b, s, h, d, attn_bias_type, attn_mask_type, dropout_probability, dtype,
-                     is_training, pad_ratio):
-        # dropout can't get the bitmatch result
-        if is_training and dropout_probability > 0.:
-            return
-
-        self.set_input(b,
-                       s,
-                       h,
-                       d,
-                       attn_bias_type=attn_bias_type,
-                       attn_mask_type=attn_mask_type,
-                       dropout_probability=dropout_probability,
-                       dtype=dtype,
-                       is_training=is_training,
-                       pad_ratio=pad_ratio)
+    def test_forward(self, b, s, h, d, attn_bias_type, attn_mask_type, backend, dropout_probability,
```

Comment:
If we have `test_forward_backward`, maybe we don't need `test_forward` anymore? Or am I misreading the code?

Reply:
I think both forward and backward needs their own unit tests so that if one of them is failed we can quickly know which fails. There is no standalone `test_backward` because the backward needs the forward output.

Reply:
Another reason is that we uses `value_and_grad` in the `test_forward_backward` for verifying the [autodiff](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html#evaluate-a-function-and-its-gradient-using-value-and-grad), which requires forward output as a scalar (like loss). So we can't verify the forward output elementwise, therefore, a standalone forward is necessary.

Reply:
I hanced the docstring of the unit test in https://github.com/NVIDIA/TransformerEngine/pull/345/commits/8d8ff5d9fdfd2e5eb12ca9ec778cc26bf0632c9e. Hope it is more descriptive.